File: packages/cli-node/src/status.ts
// Track A: `mcpseal status` — a structured, offline-first snapshot of
// local health, kept separate from the Control Plane connectivity probe
// (doctor.ts does that; status never makes a network call, so it can
// never be slow or fail because of network conditions — it answers "what
// does this machine currently believe," not "is the network up").
import path from "node:path";
import { existsSync } from "node:fs";
import { readLockfile } from "@mcpseal/cli-core";
import { readConfig } from "./config.js";
import { eventsLogPath, readEvents, recentBlocks, type McpsealEvent } from "./event-log.js";

export interface LocalStatus {
  lockfilePresent: boolean;
  lockfileError?: string;
  serverCount?: number;
  toolCount?: number;
  proxyInstalled: boolean;
  eventCount: number;
  blockCount: number;
  recentBlocks: McpsealEvent[];
}

export interface ConnectionStatus {
  loggedIn: boolean;
  workspaceId?: string;
  machineId?: string;
  ingestUrl?: string;
  lastAppliedPolicyVersion?: number;
}

export interface StatusReport {
  local: LocalStatus;
  connection: ConnectionStatus;
}

export interface StatusOptions {
  lockfilePath?: string;
  logPath?: string;
  cfgPath?: string;
}

export function buildStatusReport(projectDir: string = process.cwd(), opts: StatusOptions = {}): StatusReport {
  const lockfilePath = opts.lockfilePath ?? path.join(projectDir, ".mcp-lock.json");
  const local: LocalStatus = {
    lockfilePresent: false,
    proxyInstalled: existsSync(path.join(projectDir, ".mcp.json.mcpseal-backup")),
    eventCount: 0,
    blockCount: 0,
    recentBlocks: [],
  };

  try {
    const lockfile = readLockfile(lockfilePath);
    local.lockfilePresent = true;
    local.serverCount = Object.keys(lockfile.servers).length;
    local.toolCount = Object.values(lockfile.servers).reduce((n, s) => n + Object.keys(s.tools).length, 0);
  } catch (err) {
    local.lockfilePresent = false;
    local.lockfileError = (err as Error).message;
  }

  const events = readEvents(opts.logPath);
  local.eventCount = events.length;
  const blocks = recentBlocks(events, 10);
  local.blockCount = events.filter((e) => e.type.startsWith("blocked")).length;
  local.recentBlocks = blocks;

  const config = readConfig(opts.cfgPath);
  const connection: ConnectionStatus = {
    loggedIn: Boolean(config?.workspaceId && config?.machineId),
    workspaceId: config?.workspaceId,
    machineId: config?.machineId,
    ingestUrl: config?.ingestUrl,
    lastAppliedPolicyVersion: config?.lastAppliedPolicyVersion,
  };

  return { local, connection };
}

export function formatStatusReport(report: StatusReport): string {
  const lines: string[] = [];
  lines.push("LOCAL HEALTH");
  if (report.local.lockfilePresent) {
    lines.push(`  lockfile:  present (${report.local.serverCount} server(s), ${report.local.toolCount} tool(s) pinned)`);
  } else {
    lines.push(`  lockfile:  MISSING or invalid${report.local.lockfileError ? ` — ${report.local.lockfileError}` : ""}`);
    lines.push(`             next: mcpseal init`);
  }
  lines.push(`  proxy:     ${report.local.proxyInstalled ? "installed" : "not installed — MCP servers launch unprotected"}`);
  if (!report.local.proxyInstalled) lines.push(`             next: mcpseal install`);
  lines.push(`  events:    ${report.local.eventCount} recorded, ${report.local.blockCount} block(s) total`);
  for (const b of report.local.recentBlocks) {
    lines.push(`    [${b.ts}] ${b.type} — ${b.server}/${b.tool}`);
  }

  lines.push("");
  lines.push("CONTROL PLANE");
  if (report.connection.loggedIn) {
    lines.push(`  workspace: ${report.connection.workspaceId} (machine ${report.connection.machineId})`);
    if (report.connection.lastAppliedPolicyVersion !== undefined) {
      lines.push(`  policy:    version ${report.connection.lastAppliedPolicyVersion} last applied`);
    }
    lines.push(`  next:      mcpseal doctor   # checks live connectivity`);
  } else {
    lines.push("  not logged in — running fully local, no workspace connection.");
    lines.push("  next:      mcpseal login   # optional; local enforcement already works without it");
  }

  return lines.join("\n");
}

File: packages/cli-python/mcpseal/status.py
# Mirrors packages/cli-node/src/status.ts. Track A: `mcpseal status` — a
# structured, offline-first snapshot of local health, kept separate from
# Control Plane connectivity (doctor.py does that probe; status never
# makes a network call).
from __future__ import annotations

import os
from dataclasses import dataclass, field

from mcpseal.config import read_config
from mcpseal.event_log import events_log_path, read_events, recent_blocks
from mcpseal.lockfile import read_lockfile


@dataclass
class LocalStatus:
    lockfilePresent: bool = False
    lockfileError: str | None = None
    serverCount: int | None = None
    toolCount: int | None = None
    proxyInstalled: bool = False
    eventCount: int = 0
    blockCount: int = 0
    recentBlocks: list[dict] = field(default_factory=list)


@dataclass
class ConnectionStatus:
    loggedIn: bool = False
    workspaceId: str | None = None
    machineId: str | None = None
    ingestUrl: str | None = None
    lastAppliedPolicyVersion: int | None = None


@dataclass
class StatusReport:
    local: LocalStatus
    connection: ConnectionStatus


def build_status_report(
    project_dir: str | None = None,
    lockfile_path: str | None = None,
    log_path: str | None = None,
    cfg_path: str | None = None,
) -> StatusReport:
    resolved_project_dir = project_dir or os.getcwd()
    resolved_lockfile_path = lockfile_path or os.path.join(resolved_project_dir, ".mcp-lock.json")

    local = LocalStatus(proxyInstalled=os.path.exists(os.path.join(resolved_project_dir, ".mcp.json.mcpseal-backup")))

    try:
        lockfile = read_lockfile(resolved_lockfile_path)
        local.lockfilePresent = True
        local.serverCount = len(lockfile["servers"])
        local.toolCount = sum(len(s["tools"]) for s in lockfile["servers"].values())
    except ValueError as err:
        local.lockfilePresent = False
        local.lockfileError = str(err)

    events = read_events(log_path)
    local.eventCount = len(events)
    local.blockCount = sum(1 for e in events if e["type"].startswith("blocked"))
    local.recentBlocks = recent_blocks(events, 10)

    config = read_config(cfg_path)
    connection = ConnectionStatus(
        loggedIn=bool(config and config.get("workspaceId") and config.get("machineId")),
        workspaceId=config.get("workspaceId") if config else None,
        machineId=config.get("machineId") if config else None,
        ingestUrl=config.get("ingestUrl") if config else None,
        lastAppliedPolicyVersion=config.get("lastAppliedPolicyVersion") if config else None,
    )

    return StatusReport(local=local, connection=connection)


def format_status_report(report: StatusReport) -> str:
    lines = ["LOCAL HEALTH"]
    if report.local.lockfilePresent:
        lines.append(f"  lockfile:  present ({report.local.serverCount} server(s), {report.local.toolCount} tool(s) pinned)")
    else:
        suffix = f" — {report.local.lockfileError}" if report.local.lockfileError else ""
        lines.append(f"  lockfile:  MISSING or invalid{suffix}")
        lines.append("             next: mcpseal init")
    lines.append(f"  proxy:     {'installed' if report.local.proxyInstalled else 'not installed — MCP servers launch unprotected'}")
    if not report.local.proxyInstalled:
        lines.append("             next: mcpseal install")
    lines.append(f"  events:    {report.local.eventCount} recorded, {report.local.blockCount} block(s) total")
    for b in report.local.recentBlocks:
        lines.append(f"    [{b['ts']}] {b['type']} — {b['server']}/{b['tool']}")

    lines.append("")
    lines.append("CONTROL PLANE")
    if report.connection.loggedIn:
        lines.append(f"  workspace: {report.connection.workspaceId} (machine {report.connection.machineId})")
        if report.connection.lastAppliedPolicyVersion is not None:
            lines.append(f"  policy:    version {report.connection.lastAppliedPolicyVersion} last applied")
        lines.append("  next:      mcpseal doctor   # checks live connectivity")
    else:
        lines.append("  not logged in — running fully local, no workspace connection.")
        lines.append("  next:      mcpseal login   # optional; local enforcement already works without it")

    return "\n".join(lines)
